nextest-filtering/src/error.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) The nextest Contributors
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use miette::{Diagnostic, SourceSpan};
+use nom_tracable::TracableInfo;
+use std::cell::RefCell;
+use thiserror::Error;
+
+pub trait ToSourceSpane {
+    fn to_span(&self) -> SourceSpan;
+}
+
+#[derive(Debug, Error, Diagnostic, PartialEq, Eq)]
+pub enum Error {
+    #[error("Invalid regex")]
+    InvalidRegex(#[label("Invalid regex")] SourceSpan),
+    #[error("Expected close regex")]
+    ExpectedCloseRegex(#[label("Missing '/'")] SourceSpan),
+    #[error("Invalid identifier")]
+    InvalidIdentifier(
+        #[label("Identifier can only contain xid_continue characters and ':'")] SourceSpan,
+    ),
+    #[error("Expected identifier")]
+    ExpectedIdentifier(#[label("Missing identifier")] SourceSpan),
+    #[error("Unexpected name matcher")]
+    UnexpectedNameMatcher(#[label("This set doesn't take en argument")] SourceSpan),
+    #[error("Expected open parentheses")]
+    ExpectedOpenParenthesis(#[label("Missing '('")] SourceSpan),
+    #[error("Expected close parentheses")]
+    ExpectedCloseParenthesis(#[label("Missing ')'")] SourceSpan),
+    #[error("Expected filtering expression")]
+    ExpectedExpr(#[label("Missing expression")] SourceSpan),
+    #[error("Expected end of expression")]
+    ExpectedEndOfExpression(#[label("Unparsed input")] SourceSpan),
+
+    #[error("Unknown parsing error")]
+    Unknown,
+}
+
+#[derive(Debug, Clone)]
+pub struct State<'a>(&'a RefCell<Vec<Error>>, nom_tracable::TracableInfo);
+
+impl<'a> State<'a> {
+    pub fn new(vec: &'a RefCell<Vec<Error>>) -> Self {
+        let info = nom_tracable::TracableInfo::new()
+            .forward(true)
+            .backward(true);
+        Self(vec, info)
+    }
+
+    pub fn report_error(&self, error: Error) {
+        self.0.borrow_mut().push(error);
+    }
+}
+
+impl<'a> nom_tracable::HasTracableInfo for State<'a> {
+    fn get_tracable_info(&self) -> TracableInfo {
+        self.1.get_tracable_info()
+    }
+
+    fn set_tracable_info(mut self, info: TracableInfo) -> Self {
+        self.1 = self.1.set_tracable_info(info);
+        self
+    }
+}
+
+#[derive(Debug)]
+pub struct FilteringExprParsingError(pub String);
```

Comment:
Is there a way to make this an `ExpectedError` in the cargo-nextest crate? That will make it be printed as a nicer error (since this is a user error and not an internal error within nextest).